### PR TITLE
Stop calling read(...) if we were not able to fill the whole buffer.

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

If we not filled the whole buffer with read(...) we should stop reading and wait until we get notified again. Otherwise chances are good that the next read(...) call will either read nothing or only a very small amount of data.
Also this will allow us to call fireChannelReadComplete() which may give the user the chance to flush out all pending writes.

Modifications:

Stop reading and wait for next wakup when we not fill the whole buffer.

Result:

Less wasted read calls and early chance of flushing out pending data.

### CODE REVIEW

**Comment by normanmaurer** [comment]
This also gives a small perf improvement for HTTP benchmarks.

File: Sources/NIO/SocketChannel.swift [hunk 1]
```diff
@@ -136,12 +136,19 @@ final class SocketChannel: BaseSocketChannel<Socket> {
 
                     assert(self.isActive)
                     pipeline.fireChannelRead0(NIOAny(buffer))
-                    if mayGrow && i < maxMessagesPerRead {
+                    result = .some
+
+                    if buffer.writableBytes > 0 {
+                        // If we not filled the whole buffer with read(...) we should stop reading and wait until we get notified again.
```
Comment (weissi) [inline Sources/NIO/SocketChannel.swift]: grammar nit: 'If we did not fill the whole buffer' ...

Comment (normanmaurer) [reply Sources/NIO/SocketChannel.swift]: fixed

**Comment by normanmaurer** [comment]
@weissi never! Fixed it 👍

File: Tests/NIOTests/ChannelTests.swift [hunk 2]
```diff
@@ -1963,12 +1972,14 @@ public class ChannelTests: XCTestCase {
...
-                XCTFail("channelReadComplete unexpected")
+                if self.state != .read {
+                    XCTFail("channelReadComplete unexpected")
+                }
                 self.state = .readComplete
             }
 
             func errorCaught(ctx: ChannelHandlerContext, error: Error) {
-                XCTAssertEqual(.read, self.state)
+                XCTAssertTrue(.read == self.state || .readComplete == self.state)
```
[the remark this answers is not in the document]
Comment (normanmaurer) [reply Tests/NIOTests/ChannelTests.swift]: actually I think  `.readComplete ==` is the only one we need to test... let me fix

**Comment by normanmaurer** [comment]
@Lukasa addressed

**Comment by Lukasa** [comment]
Rebase and merge when ready.

**Pull request merged**